scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -973,16 +973,61 @@
                                        {:job-uuid->reserved-host (apply dissoc job-uuid->reserved-host matched-job-uuids)
                                         :launched-job-uuids (into matched-job-uuids launched-job-uuids)})))
 
+(defn job->preferred-compute-clusters
+  "Given a job and a collection of compute clusters, returns the
+  subset of compute clusters that the job would prefer to run
+  (and therefore, autoscale) on"
+  [{:keys [job/checkpoint job/instance]} compute-clusters]
+  (if (and checkpoint instance)
+    ; If checkpointing is enabled, we want to run the new instance in the
+    ; same location as the checkpointed instance to take advantage of data
+    ; locality for the checkpoint data
+    (let [{{:keys [compute-cluster/cluster-name]} :instance/compute-cluster}
+          (->> instance
+               (sort-by :instance/start-time)
+               last)
+          ; This is a list, but it should never
+          ; contain more than 1 compute cluster
+          previous-compute-clusters
+          (filter
+            (fn [{:keys [name]}]
+              (= name cluster-name))
+            compute-clusters)]
+      (if (seq previous-compute-clusters)
+        (let [previous-location
+              (-> previous-compute-clusters
+                  first
+                  :cluster-definition
+                  :config
+                  :location)]
+          (filter
+            (fn [{{{:keys [location]} :config} :cluster-definition}]
+              (= location previous-location))
+            compute-clusters))
+        ; If the previous instance's compute cluster name is not
```

Comment:
if we ever don't match the location we want here, what will happen when we try to match the offer?
will we launch synthetic pods but then never match any offers?

Reply:
The matching is going to have to be implemented as a "yes"/"no" constraint. So if there are no offers in the same location as the previous instance, then we won't be able to match the job to anything.